docs: add contributing.md (#941)
coveo/cli: CONTRIBUTING.md

# Contributing to the Coveo CLI

The Coveo CLI repository is organized as a monorepo. To learn more about repository, check-out [monorepo.tools](https://monorepo.tools/).

> Its packages are split into 3 big categories:

Reviewer: ```suggestion
Its packages are split into three categories:
```

> - UI: Under `packages/ui` you'll find the project 'gravitating' around the CLI that helps developers to get started with a Headless implementation.

Reviewer: ```suggestion
- UI: Under `packages/ui` you'll find the project 'gravitating' around the CLI that helps developers get started with a Headless implementation.
```

> - UI: Under `packages/ui` you'll find the project 'graviting' around the CLI and that helps developer to get started with a Headless implementation.

Author: It's not in master yet it'll be done in #939 
This is why this PR is actually a follow-up on #939

- CLI: Under `packages/cli` you'll find the 'vanilla' plugin of the CLI, the CLI itself, and some utils to help with the development of the CLI and its plugins.
<!-- TODO: Remove when the E2E will finally be 'split' -->

> - E2E: A special 'package' under `packages/cli-e2e`, it contains the end-to-end tests for the CLI and its included plugin.

Reviewer: ```suggestion
- E2E: A special 'package' under `packages/cli-e2e`, which contains the end-to-end tests for the CLI and its included plugins.
```

The rest of this document will focus on the CLI and its plugins.

## The CLI folder

> The CLI folder contains 3 special projects:

Reviewer: ```suggestion
The CLI folder contains three special projects:
```

- `cli-commons-dev`: a set of development utilities (e.g. eslint config) shared across the CLI and its plugins.
- `cli-commons`: a set of **production** utilities (e.g. error wrapping) shared across the CLI and its plugins.
- `core`: The CLI Itself.

This structure allows the CLI to ship as a bundle of bricks, rather than as a big chunky house.

> However, our tools don't necessarily play nicely with this structure.

Author: ```suggestion
However, `oclif` doesn't necessarily play nicely with this project structure
```

> However, our tools doesn't play necessarly nicely with this structure.

Author: I will clarify, but essentially: Oclif. It required quite a few tweaks

### The TypeScript Configs

You'll find up to 3 typescript configs in the CLI projects:

- `tsconfig.json`:
  - Here to serve as a 'local' base configuration for the other, but more importantly
  - Used as default by the language server of your IDE (think wiggly red lines)
- `tsconfig.dev.json`:

>   - Only for CLI and plugins, it is used when you run the commands in **dev** mode with OClif

Reviewer: ```suggestion
  - Only for CLI and plugins. Used when you run the commands in **dev** mode with OClif.
```

  - To work, it needs:
    - The `@oclif/core` patch
    - The custom `dev` node executable
- `tsconfig.build.json`:
  - Only used for the transpilation of the shipped packages.
  - It uses the aforementioned references to identify the 'external' bits.

### The package.json Files

> When checking `cli-commons` and `cli-commons-dev`, you'll see an `exports` field.

Reviewer: ```suggestion
When looking at `cli-commons` and `cli-commons-dev`, you'll see an `exports` field.
```

This give instruction on where to find some files to the package resolver, so that the import interface stay simples and consistent between development and production.

For more info on that, check [NodeJS documentation](https://nodejs.org/api/packages.html#subpath-exports)

> All those configurations, one way or another, does point to the `tsconfig.json` of `cli-commons-dev`.

Reviewer: oups I just realized that I may have edited some of Yassine's suggestions by mistake 🙊 
Hopefully I won't have caused confusion 🙃
